python/ray/rllib/utils/torch_policy_graph.py

```diff
@@ -0,0 +1,106 @@
+
+    This is similar to TFPolicyGraph, but for PyTorch.
+
+    Attributes:
+        observation_space (gym.Space): observation space of the policy.
+        action_space (gym.Space): action space of the policy.
+        lock (Lock): Lock that must be held around PyTorch ops on this graph.
+    """
+
+    def __init__(
+            self, observation_space, action_space, model, loss, loss_inputs):
+        """Initialize the policy graph.
+
+        Arguments:
+            observation_space (gym.Space): observation space of the policy.
+            action_space (gym.Space): action space of the policy.
+            model (nn.Module): PyTorch policy module. Given observations as
+                input, this module must a list of outputs where the first item
+                are action logits, and the remainder can be any value.
+            loss (nn.Module): Loss defined as a PyTorch module. The inputs for
+                this module are defined by the `loss_inputs` param. This module
+                returns a single scalar loss.
+            loss_inputs (list): List of SampleBatch columns that will be
+                passed to the loss module's forward() function when computing
+                the loss. For example, ["obs", "action", "advantages"].
+
+        Attributes:
+            observation_space (gym.Space): observation space of the policy.
+            action_space (gym.Space): action space of the policy.
+        """
+        self.observation_space = observation_space
+        self.action_space = action_space
+        self.lock = Lock()
+        self._model = model
+        self._loss = loss
+        self._loss_inputs = loss_inputs
+        self._optimizer = self.optimizer()
+
+    def extra_action_out(self, model_out):
+        """Returns dict of extra info to include in experience batch.
+
+        Arguments:
+            model_out (list): Outputs of the policy model module."""
+        return {}
+
+    def optimizer(self):
+        """Custom PyTorch optimizer to use."""
+        return torch.optim.Adam(self._model.parameters())
+
+    def compute_actions(
+            self, obs_batch, state_batches=None, is_training=False):
+        if state_batches:
+            raise NotImplementedError("Torch RNN support")
+        with self.lock:
+            ob = torch.from_numpy(np.array(obs_batch)).float()
+            model_out = self._model(ob)
+            logits = model_out[0]  # assume the first output is the logits
+            actions = F.softmax(logits, dim=1).multinomial(1).squeeze(0)
+            return var_to_np(actions), [], self.extra_action_out(model_out)
+
+    def compute_gradients(self, postprocessed_batch):
+        with self.lock:
+            loss_in = []
+            for key in self._loss_inputs:
+                loss_in.append(torch.from_numpy(postprocessed_batch[key]))
+            loss_in = [
+                t.float() if isinstance(t, torch.DoubleTensor) else t
```

Comment:
looks brittle

Reply:
Do you have a better idea? Ideally I'd insert an implicit cast to the right tensor type here.

Reply:
sounds like it should be on the user to make the right numpy type, but I can see this doubletensor thing being a common bug. what do you do for TF?